fv3core/stencils/d_sw.py

```diff
@@ -93,8 +94,92 @@ def horizontal_relative_vorticity_from_winds(
         vorticity[0, 0, 0] = rarea * (vt - vt[0, 1, 0] - ut + ut[1, 0, 0])
 
 
+@gtscript.function
+def add_dw(w, dw, damp_w):
+    w = w + dw if damp_w > 1e-5 else w
+    return w
+
+
+@gtscript.function
+def ke_from_bwind(ke, ub, vb):
+    return 0.5 * (ke + ub * vb)
+
+
+@gtscript.function
+def adjust_w_and_qcon(w, delp, dw, damp_w, q_con):
+    w = w / delp
+    w = add_dw(w, dw, damp_w)
+    q_con = q_con / delp
+
+    return w, q_con
+
+
 @gtstencil()
-def not_inlineq_pressure(gx: sd, gy: sd, rarea: sd, fx: sd, fy: sd, pt: sd, delp: sd):
+def ke_horizontal_vorticity_w_qcon_adjust(
+    ke: FloatField,
+    u: FloatField,
+    v: FloatField,
+    ub: FloatField,
+    vb: FloatField,
+    ut: FloatField,
+    vt: FloatField,
+    dx: FloatField,
+    dy: FloatField,
+    rarea: FloatField,
+    vorticity: FloatField,
+    w: FloatField,
+    delp: FloatField,
+    dw: FloatField,
+    q_con: FloatField,
+    dt: float,
+    nested: bool,
+    damp_w: float,
+):
+    from __externals__ import (
+        i_end,
+        i_start,
+        j_end,
+        j_start,
+        local_ie,
+        local_is,
+        local_je,
+        local_js,
+    )
+
+    with computation(PARALLEL), interval(...):
+        ke = ke_from_bwind(ke, ub, vb)
+
+        with horizontal(region[i_start, j_start]):
+            if not nested:
+                ke = corners.corner_ke(ke, u, v, ut, vt, dt, 0, 0, -1, 1)
+        with horizontal(region[i_end + 1, j_start]):
+            if not nested:
+                ke = corners.corner_ke(ke, u, v, ut, vt, dt, -1, 0, 0, -1)
+        with horizontal(region[i_end + 1, j_end + 1]):
+            if not nested:
+                ke = corners.corner_ke(ke, u, v, ut, vt, dt, -1, -1, 0, 1)
+        with horizontal(region[i_start, j_end + 1]):
+            if not nested:
+                ke = corners.corner_ke(ke, u, v, ut, vt, dt, 0, -1, -1, -1)
```

Comment:
Awesome! So you probably weren't aware -- but 'with horizontal' can happen inside of functions, which is really handy. So you could put all of these regions into a single gtscript function named something like corners.all_corners_ke (and still reuse your corner_ke function from within it!).   You will need to restate 'from__externals__ ...'

Also, it's great you kept the nested logic, but I think we can not include it as long as we document it somewhere. c_sw there is a comment " # additional assumption: not __INLINED(spec.grid.nested)" that we probably want to add to the ke corners function. Then you don't have to pass in nested at all and don't need as many if statements.

Reply:
Done!  I created a function `all_corners_ke` that encompasses all the calls to `corners.corner_ke` using `with_horizontal`...